### PR TITLE
Torch backend `aten::cat.out` fix

### PR DESCRIPTION
**Pull request by VikramRangarajan**
**Note:** This is a (hopefully) fixed version of #16051, I was told the code wasn't ready to merge. I have added more tests, cleaned up the code, and added support for negative `dim`. I also fixed the function signature for `aten::cat.out` according to https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L1438. Please let me know if there are further problems with this PR.

## Motivation

tinygrad.Tensor.cat requires: All tensors must have the same shape except in the concatenating dimension.

torch.cat requires: All tensors must either have the same shape (except in the concatenating dimension) or be a 1-D empty tensor with size (0,).

This adds the 1-D empty tensor fix by reshaping the empty tensor to the output shape, but with a 0 in the dimension you are cat'ing over. This makes it compatible with tinygrad's cat, it's zero copy (reshape with empty tensor), and preserves the graph (we don't exclude the empty tensors from the computation).

Example:

Before:

```
import torch
from extra.torch_backend import backend
torch.cat((torch.empty([2, 2], device="tiny"), torch.empty([0], device="tiny")), dim=0) # error (works on cpu)
```

Now, it takes the first shape (0,) tensor and reshapes it into (0, 2), then does tinygrad.Tensor.cat with a (0, 2) and (2, 2) tensor, which works.

A practical example of this edge case being used is huggingface transformers KV cache: https://github.com/huggingface/transformers/blob/main/src/transformers/cache_utils.py#L120-L145. Right now we cannot run transformers models by doing `model.to("tiny").generate(...)` as the current cat implementation doesn't support 1d empty tensors.

### CODE REVIEW

**Comment by geohot** [comment]
Are you sure you don't just want to add support for this to Tensor.cat in tinygrad? Also, be careful, do not use AI.

**Comment by VikramRangarajan** [comment]
I only used AI to figure out that the problem with model.generate was this edge case in aten::cat.out, but the fix and tests are handwritten.
For Tensor.cat in tinygrad, I feel like it wouldn't be a good change. PyTorch is the only library that has this weird empty empty 1d allowance, numpy, jax, and the array api, all agree with tinygrad's current cat, so I think it makes sense for this to stay in the torch backend.
